### PR TITLE
571 aliens on available nodes

### LINKED ISSUE #571: Take nodes availability into account when choosing support nodes for aliens
Current implementation does not check nodes availability status and therefore leads to a decrease in efficiency and reliability.

https://github.com/qoollo/bob/blob/eab98a3cd1614b1eae1df20809559bbfb3f20573/bob/src/cluster/quorum.rs#L142
https://github.com/qoollo/bob/blob/ab2b97c32fed59fe31aad50c1610ab3891777a5c/bob-common/src/mapper.rs#L118

### PR DESCRIPTION
**Pull request by idruzhitskiy**
Closes #571

### CODE REVIEW

File: bob-common/src/node.rs
Comment (ikopylov) [inline bob-common/src/node.rs:90]: Use `Acquire` + `Release` ordering

File: bob-common/src/node.rs
Comment (ikopylov) [inline bob-common/src/node.rs:77]: Do this update inside `self.conn.write()` lock. Otherwise the following situation is possible:
1. Thread1: `set_connection` is called and update connection
2. Thread2: `clear_connection` is called and reset connection
3. Thread2: store `false` to `conn_available`
4. Thread1: store `true` to `conn_available`

File: bob-common/src/node.rs
Comment (ikopylov) [inline bob-common/src/node.rs:82]: Do this update inside `write` lock

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:145]: How does `contains` work for `Node` struct? Does it compare the full struct content? If this is the true, then it is inefficient. It is better to look for the presence of a node by comparing only its `index`

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:139]: This heavily overlaps with https://github.com/qoollo/bob/pull/629. Once #629 will be merged, this PR should be updated

File: CHANGELOG.md
Comment (ikopylov) [inline CHANGELOG.md:62]: Merge conflict not fixed properly

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:137]: This is not good:
1. Iterations continue even when `support_nodes` are full. Previously there was a `break` to stop the iteration
2. This ignores the idea of uniform alien distribution, that was introduced by https://github.com/qoollo/bob/pull/629. Imagine the situation when in 4 nodes cluster, nodes 1 and 2 are tuned off. According to the implemented logic node 3 will keep 3 times more aliens than node 4. We need to make a uniform distribution, taking into account the availability of the connection

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:143]: Probably, we need a single hand written circular for loop to properly implement skipping of disconnected nodes with offset

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:58]: `async` is not required

Comment (ikopylov) [reply bob-common/src/mapper.rs:58]: And in `new` above too

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:131]: No need to use iterator with vector. It is much simpler to use indexation: `self.nodes[i]`. Thus `for i in 0..nodes.len()` or some kind of `while` loop will be better

Comment (ikopylov) [reply bob-common/src/mapper.rs:131]: Tested algo on C#:
```C#
    private (int index, int avail) Find(bool[] status, int offset)
    {
        int avail = 0;
        for (int i = 0; i < status.Length; i++)
        {
            if (status[i])
            {
                if (avail == offset)
                    return (i, avail + 1);

                avail++;
            }
        }

        return (-1, avail);
    }

    public int Select(bool[] status)
    {
        int offset = Interlocked.Increment(ref _offset);
        var res = Find(status, offset % status.Length);

        if (res.index < 0 && res.avail > 0)
        {
            int curr_offset = offset % res.avail;
            while (res.index < 0 && res.avail > 0)
            {
                res = Find(status, curr_offset);
                curr_offset = Math.Max(0, curr_offset - res.avail);
            }
        }

        if (res.index < 0)
            return offset % status.Length;

        return res.index;
    }
```

File: CHANGELOG.md
Comment (ikopylov) [inline CHANGELOG.md:17]: Move this line to `[Unreleased]` section

File: bob-common/src/data.rs
Comment (ikopylov) [inline bob-common/src/data.rs:312]: Can this be `&[Node]`?

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:133]: You forget about `target_nodes.iter().all(|n| n.index() != node.index())` condition

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:146]: `node.connection_available()` is faster to check, so it should be the first

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:140]: This is offset within available nodes. It is incorrect to use it as offset within the whole `nodes` collection

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:132]: I suggest to extract initial node search logic into a separate function

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:144]: With for cycle you can easily implement cycle iterations:
```
for _ in 0..self.nodes.len() {
    index = (index + 1) % self.nodes.len();
}
```

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:154]: This can be merged with previous loop if you use cycle iteration approach

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:155]: This can be merged with previous loop if you use cycle iteration approach

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:154]: You forget to fill `support_nodes` at the end with disconnected ones if `support_nodes.len() < count`

File: bob-common/src/mapper.rs
Comment (ikopylov) [reply bob-common/src/mapper.rs:140]: That's why I introduced additional calls to `Find` within loop in my example

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:21]: Bad name

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:124]: Expand the name to make it clear that the function is for the search for the support node

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:138]: Same here

File: bob-common/src/mapper.rs
Comment (ikopylov) [reply bob-common/src/mapper.rs:124]: It is better to make function static: don't receive `self` as parameter

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:128]: You forget about the following condition: `target_nodes.iter().all(|n| n.index() != node.index())`

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:144]: You can move this condition to the line after `res = self.find_index(curr_offset);`. This should simplify the code

Comment (ikopylov) [reply bob-common/src/mapper.rs:144]: Or you can add a function that checks that res is `Avail` and inner value is greater than zero

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:148]: Bug here. You should use current `res` value

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:179]: Why do you use this strange condition?

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:184]: Very strange condition. Proper condition:
```Rust
if node.connection_available() && support_nodes.iter().all(|&n| n.index() != node.index()) {
```

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:194]: After this cycle ends, you should check, whether the `support_nodes` list is full. If it is not, then you should add also a disconnected nodes

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:194]: Please, add comment, that here we try to fill `support_nodes` ignoring connection status

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:193]: We may not fill the list also if `nodes.len() - target_nodes.len() < count`. So I suggest to extend that condition
```suggestion
        if support_nodes.len() < count && support_nodes.len() +  target_nodes.len() < nodes.len() {
```

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:146]: Write the comment, that this function looks for an offset in the list of nodes, respecting the uniform distribution of resulting nodes

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:147]: Actually, we know, that we always skip target nodes, so we can calculate local offset as `offset % (nodes.len() - target_nodes.len())`. But to make it safe, at the begging of the function, we should explicitly check that `target_nodes.len() < nodes.len()`. If it is not, then return 0 immediately. Also in `get_support_nodes` the same check can be added as a short-cut: we can return empty vector in this case

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:124]: I don't like this name. Maybe `try_find_support_node_offset_at_one_pass`?

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:176]: Write the comment that the idea of the current logic is to preserve uniform distribution of aliens

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:171]: Please, add a `debug_assert` that checks that all nodes in `target_nodes` are unique. You can do that by collection nodes indices into HashSet and compare lengths

File: bob-common/src/mapper.rs
Comment (lostl1ght) [reply bob-common/src/mapper.rs:145]: Done

File: bob-common/src/mapper.rs
Comment (lostl1ght) [reply bob-common/src/mapper.rs]: Done.

File: bob-common/src/mapper.rs
Comment (lostl1ght) [reply bob-common/src/mapper.rs]: Done.

File: bob-common/src/mapper.rs
Comment (lostl1ght) [reply bob-common/src/mapper.rs:187]: Done.

File: bob-common/src/mapper.rs
Comment (lostl1ght) [reply bob-common/src/mapper.rs:177]: Done.

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:191]: You should pass the offset without calculating the remainder, since at this level it is necessary to abstract from the offset search logic implemented in `find_support_node_offset`

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:147]: `offset % nodes.len()` is better as fallback

File: bob-common/src/mapper.rs
Comment (ikopylov) [inline bob-common/src/mapper.rs:210]: Please, explain why the connection status is ignored (this is because we are trying to fill the support nodes vector up to the requested `count` of nodes)

File: CHANGELOG.md
Comment (ikopylov) [inline CHANGELOG.md:22]: Please, move this line to `[Unreleased]` section

**Review by ikopylov — APPROVED** [review]
LGTM

**Pull request merged**
